src/components/charcount/charcount.component.tsx

```diff
@@ -4,8 +4,8 @@ import React from 'react';
 
 export interface ICharCountProps extends IComponentProps {
   count: number;
-
   pristine: boolean;
+  showCharCount: boolean;
```

Comment:
A number input can have showCharCount right? My idea was that input that can't have a certain reusable feature (showCharCount for checkbox) will simply not implement <CharCount />. Once <CharCount /> is called all the logic can be kept in one place utilizing the consistent (as possible) interface. This prevents this kind of logic from leaking into each implementation.

Reply:
I don't see it in vanilla formio as a builder option: https://formio.github.io/formio.js/app/builder.html

And I'm 100% certain that we don't expose that option in our own builder configuration.

> This prevents this kind of logic from leaking into each implementation.

I do get this, but some (seemingly) re-usable components are actually very situational and different implementations may even deliberately choose to (not) support them, which is a valid choice. I can perfectly imagine that a webcomponent textfield or textarea implementation already has a similar feature implemented themselves.

At this time I'm really leaning towards letting the formio component type components (textfield, textarea, number...) do their own composition instead of us doing the "generic" composition. Because if we do it, then other implementations have no way to *not* use our wrappers except for overriding the wrapper component itself, which introduces a whole bunch of `if component.type` checks that they may not want to deal with.

For example, textfield, textarea, select etc. all likely need the outer `div` wrapper, but a checkbox component has very different markup where this outer wrapper div is not relevant and should not be present.

IMO the primary configuration interface should be the mapping of formio component type to the component to use. While being able to override the `label` / `description` etc. components, I do consider those a secondary layer that's further removed from public API and closer to private-ish API with more risks (i.e.: know what you're doing and be ready to deal with unpleasant changes between versions).

Reply:
I'm not pointing to third-party (web) components here, but rather to our own components which will all benefit from a clearer separation. I agree that if we choose not to support certain features for subtypes (e.g.) number that those edge cases should be implemented in the components itself (e.g <TextField/>. But IMO the generic behavior should be implemented where it is used, the <CharCount/> itself.

I see an overlap with errors which isolates the display logic (whether the error tree should be rendered at all) as much as possible only adding a minimal additional interface (pristine) which should be derived from the components/form state.

By isolating such decision making a component can be much cleaner and risk of inconsistent behavior can be reduced.

Third party components have to take care of their own implementation. I don't think we need to expose partials to the outside world.